client.go

```diff
@@ -1,48 +1,273 @@
-	return NewWithClient(amzCli, bucketName), nil
+	cli := Instantiate(sdkClient, cryptoClient, bucketName, region)
+	return cli, nil
 }
 
-// NewWithClient returns a new S3 structure for the provided AmzClient instance.
-func NewWithClient(client AmzClient, bucketName string) *S3 {
+// Instantiate creates a new instance of S3 struct with the provided clients, bucket and region
+func Instantiate(sdkClient S3SDKClient, cryptoClient S3CryptoClient, bucketName, region string) *S3 {
 	return &S3{
-		cli:        client,
-		BucketName: bucketName,
+		sdkClient:    sdkClient,
+		cryptoClient: cryptoClient,
+		bucketName:   bucketName,
+		region:       region,
 	}
 }
 
-// Get returns an io.ReadCloser instance for the given fully qualified S3 URL.
-// Note that this function will mutate the bucket name of the S3 object if it's different than the existing one.
-func (s3 *S3) Get(rawURL string) (io.ReadCloser, error) {
+// BucketName is a getter for the bucket name used by this S3 client
+func (cli *S3) BucketName() string {
+	return cli.bucketName
+}
+
+// Upload handles the uploading a file to AWS S3, into the bucket configured for this client
+func (cli *S3) Upload(ctx context.Context, req *UploadRequest, payload []byte) error {
+	return cli.UploadWithPsk(ctx, req, payload, nil)
+}
+
+// UploadWithPsk handles the uploading a file to AWS S3, into the bucket configured for this client, using a user-defined psk
+func (cli *S3) UploadWithPsk(ctx context.Context, req *UploadRequest, payload []byte, psk []byte) error {
+
+	listMultiInput := &s3.ListMultipartUploadsInput{
+		Bucket: &cli.bucketName,
+	}
+
+	listMultiOutput, err := cli.sdkClient.ListMultipartUploads(listMultiInput)
+	if err != nil {
+		log.Event(ctx, "error fetching multipart list", log.Error(err))
+		return err
+	}
+
+	var uploadID string
+	for _, upload := range listMultiOutput.Uploads {
+		if *upload.Key == req.UploadKey {
+			uploadID = *upload.UploadId
+		}
+	}
+
+	if len(uploadID) == 0 {
+
+		createMultiInput := &s3.CreateMultipartUploadInput{
+			Bucket:      &cli.bucketName,
+			Key:         &req.UploadKey,
+			ContentType: &req.Type,
+		}
+
+		createMultiOutput, err := cli.sdkClient.CreateMultipartUpload(createMultiInput)
+		if err != nil {
+			log.Event(ctx, "error creating multipart upload", log.Error(err))
+			return err
+		}
+
+		uploadID = *createMultiOutput.UploadId
+	}
+
+	rs := bytes.NewReader(payload)
+
+	n := int64(req.ChunkNumber)
+
+	uploadPartInput := &s3.UploadPartInput{
+		UploadId:   &uploadID,
+		Bucket:     &cli.bucketName,
+		Key:        &req.UploadKey,
+		Body:       rs,
+		PartNumber: &n,
+	}
+
+	if psk != nil {
+		_, err = cli.cryptoClient.UploadPartWithPSK(uploadPartInput, psk)
+		if err != nil {
+			log.Event(ctx, "error uploading with psk", log.Error(err))
+			return err
+		}
+	} else {
+		_, err = cli.sdkClient.UploadPart(uploadPartInput)
+		if err != nil {
+			log.Event(ctx, "error uploading part", log.Error(err))
+			return err
+		}
+	}
```

Comment:
Consider splitting this out into a separate func something like `func doUploadPart(part, psk) error`